### PR TITLE
add FailedToRetrieveImagePullSecret event.

### LINKED ISSUE #104432: imagePullSecrets should log warning if secret does not exist
#### What happened:

imagePullSecrets can added to a pod manifest to reference a secret containing registry credentials. Currently, if you add the name of a secret that does not exist, there is no warning or error. 

#### What you expected to happen:

I would expect there to be an error and a failed deployment if the secret does not exist/cannot be found in the namespace. 
If people are pulling public images, they may still wish to authenticate to a registry, for example dockerhub, before pulling the image, in order to utilise their subscription account rate limits. Currently if they use imagePullSecrets and the referenced secret does not exist, they may assume they are authenticating to their account when they are not, and will still be able to pull public images without any warning that authentication is not taking place. 

#### How to reproduce it (as minimally and precisely as possible):

Add `imagePullSecrets: myfakesecret` to a pod spec, and pull images from docker.io.

### PR DESCRIPTION
**Pull request by calvin0327**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
We want to add a warning event if imagePullSecrets does not exist.

#### Which issue(s) this PR fixes:

Fixes #104432

#### Special notes for your reviewer:
This is my good-first-issue, please indicate my mistake.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by matthyx** [comment]
> Thanks for your pull request. Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).
> 
>  **Please follow instructions at https://git.k8s.io/community/CLA.md#the-contributor-license-agreement to sign the CLA.**
> 
> It may take a couple minutes for the CLA signature to be fully registered; after that, please reply here with a new comment and we'll verify. Thanks.
> 
> * If you've already signed a CLA, it's possible we don't have your GitHub username or you're using a different email address.  Check your existing CLA data and verify that your [email is set on your git commits](https://help.github.com/articles/setting-your-email-in-git/).
> * If you signed the CLA as a corporation, please sign in with your organization's credentials at https://identity.linuxfoundation.org/projects/cncf to be authorized.
> * If you have done the above and are still having issues with the CLA being reported as unsigned, please log a ticket with the Linux Foundation Helpdesk: https://support.linuxfoundation.org/
> * Should you encounter any issues with the Linux Foundation Helpdesk, send a message to the backup e-mail support address at: [login-issues@jira.linuxfoundation.org](https://mail.google.com/mail/?view=cm&fs=1&tf=1&to=login-issues@jira.linuxfoundation.org)

@calvin0327 you need to sign the CLA

File: pkg/kubelet/kubelet_pods.go [hunk 1]
```diff
@@ -897,6 +897,10 @@ func (kl *Kubelet) getPullSecretsForPod(pod *v1.Pod) []v1.Secret {
 		pullSecrets = append(pullSecrets, *secret)
 	}
 
+	if len(pullSecrets) == 0 {
+		kl.recorder.Eventf(pod, v1.EventTypeWarning, "FailedToRetrieveImagePullSecret", "Unable to retrieve pull secret, the image pull may not succeed.")
```
Comment (adisky) [inline pkg/kubelet/kubelet_pods.go:901]: I would record the event after L-891, here  if there are two pull secret and one is success and other is failed, then we will miss event for one
```
if err != nil {
    klog.InfoS("Unable to retrieve pull secret, the image pull may not succeed.", "pod", klog.KObj(pod), "secret", klog.KObj(secret), "err", err)
    if errors.IsNotFound(err) {
        kl.recorder.Eventf(pod, v1.EventTypeWarning, "FailedToRetrieveImagePullSecret", "Unable to retrieve image pull secret %s, the image pull may not succeed.", secretRef.Name)
     }
     continue
  }
```

Comment (kerthcet) [reply pkg/kubelet/kubelet_pods.go:901]: To ensure the warning event  be emitted exactly once, maybe we can achieve like this: 
```
var  secretNotExists bool

for _, secretRef := range pod.Spec.ImagePullSecrets {
    secret, err := kl.secretManager.GetSecret(pod.Namespace, secretRef.Name)
    if err != nil {
	    klog.InfoS("Unable to retrieve pull secret, the image pull may not succeed.", "pod", klog.KObj(pod), 
    "secret", klog.KObj(secret), "err", err)
            secretNotExists = true
    }
}

if secretNotExists {
    kl.recorder.Eventf()
}
```

Comment (calvin0327) [reply pkg/kubelet/kubelet_pods.go:901]: I think the warning event should be generated once in the function, correct me if not.

Comment (adisky) [reply pkg/kubelet/kubelet_pods.go:901]: right, event should be generated once in the function, how about this?
```
var failedsecrets  []string
if err != nil {
    klog.InfoS("Unable to retrieve pull secret, the image pull may not succeed.", "pod", klog.KObj(pod), "secret", klog.KObj(secret), "err", err)
    failedsecrets = append(failedsecrets,secretRef.Name )
    continue
  }

if len(failedsecrets) != 0 {
    kl.recorder.Eventf(pod, v1.EventTypeWarning, "FailedToRetrieveImagePullSecret", "Unable to retrieve image pull secret %s, the image pull may not succeed.", failedsecrets)
}
```

Comment (kerthcet) [reply pkg/kubelet/kubelet_pods.go:901]: lgtm

File: pkg/kubelet/kubelet_pods.go [hunk 2]
```diff
@@ -891,12 +892,17 @@ func (kl *Kubelet) getPullSecretsForPod(pod *v1.Pod) []v1.Secret {
...
+			failedSecrets = append(failedSecrets, secretRef.Name)
 			continue
 		}
 
 		pullSecrets = append(pullSecrets, *secret)
 	}
 
+	if len(failedSecrets) != 0 {
+		kl.recorder.Eventf(pod, v1.EventTypeWarning, "FailedToRetrieveImagePullSecret", "Unable to retrieve image pull secret %s, the image pull may not succeed.", failedSecrets)
+	}
```
Comment (kerthcet) [inline pkg/kubelet/kubelet_pods.go:904]: as `failedSecrets` is a slice type, the `Eventf` output may looks like `image pull secret [secretA, secretB], the image...`. We can refactor with `strings.Join(failedSecrets, ", ")`. I think it's prettier.

Comment (calvin0327) [reply pkg/kubelet/kubelet_pods.go:904]: Yes, I was careless.

Comment (kerthcet) [reply pkg/kubelet/kubelet_pods.go:904]: Another advice, rebase multi commits to one commit as they are doing the same thing, we can keep the baseline of master branch clearly.

**Review by adisky — COMMENTED** [review]
/lgtm
/assign @SergeyKanzhelev @ehashman

**Comment by calvin0327** [comment]
/assign @mrunalp
for approval

File: pkg/kubelet/kubelet_pods.go [hunk 3]
```diff
@@ -891,12 +892,17 @@ func (kl *Kubelet) getPullSecretsForPod(pod *v1.Pod) []v1.Secret {
 		secret, err := kl.secretManager.GetSecret(pod.Namespace, secretRef.Name)
 		if err != nil {
 			klog.InfoS("Unable to retrieve pull secret, the image pull may not succeed.", "pod", klog.KObj(pod), "secret", klog.KObj(secret), "err", err)
+			failedSecrets = append(failedSecrets, secretRef.Name)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet_pods.go:895]: let's include the namespace into this string

File: pkg/kubelet/kubelet_pods.go [hunk 4]
```diff
@@ -891,12 +892,17 @@ func (kl *Kubelet) getPullSecretsForPod(pod *v1.Pod) []v1.Secret {
...
 			continue
 		}
 
 		pullSecrets = append(pullSecrets, *secret)
 	}
 
+	if len(failedSecrets) != 0 {
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet_pods.go:902]: please add unit test for this. There seems to already be the fakeRecorder example in the corresponding test file

File: pkg/kubelet/kubelet_pods.go [hunk 5]
```diff
@@ -891,12 +892,17 @@ func (kl *Kubelet) getPullSecretsForPod(pod *v1.Pod) []v1.Secret {
...
+		kl.recorder.Eventf(pod, v1.EventTypeWarning, "FailedToRetrieveImagePullSecret", "Unable to retrieve image pull secrets %s, the image pull may not succeed.", strings.Join(failedSecrets, ", "))
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet_pods.go:903]:
```suggestion
		kl.recorder.Eventf(pod, v1.EventTypeWarning, "FailedToRetrieveImagePullSecret", "Unable to retrieve image pull secrets %s, will attempt to pull the image without these secrets. The image pull may not succeed.", strings.Join(failedSecrets, ", "))
```

**Review by SergeyKanzhelev — COMMENTED** [review]
it would be great to have a test for it

**Comment by SergeyKanzhelev** [comment]
/lgtm cancel

my biggest concern is that namespace is not added to the secret name. Tests would be very nice to have.

**Comment by calvin0327** [comment]
@SergeyKanzhelev Thanks for review, Can you tell me where the fakeRecorder example is? I looked for it for a long time.

**Comment by haorenfsa** [comment]
> @SergeyKanzhelev Thanks for review, Can you tell me where the fakeRecorder example is? I looked for it for a long time.

@calvin0327 see [pkg/kubelet/kubelet_pods_test.go: line 1672](https://github.com/kubernetes/kubernetes/blob/823cc3cc3671acca94ae16cb7e93a4857a4526eb/pkg/kubelet/kubelet_pods_test.go#L1672)

**Comment by calvin0327** [comment]
> > @SergeyKanzhelev Thanks for review, Can you tell me where the fakeRecorder example is? I looked for it for a long time.
> 
> @calvin0327 see [pkg/kubelet/kubelet_pods_test.go: line 1672](https://github.com/kubernetes/kubernetes/blob/823cc3cc3671acca94ae16cb7e93a4857a4526eb/pkg/kubelet/kubelet_pods_test.go#L1672)

**Comment by dims** [comment]
@calvin0327 there are some suggestions pending and a CI job failure as well. will you be revisiting this PR?

File: pkg/kubelet/kubelet_pods.go [hunk 6]
```diff
@@ -891,12 +892,17 @@ func (kl *Kubelet) getPullSecretsForPod(pod *v1.Pod) []v1.Secret {
 		secret, err := kl.secretManager.GetSecret(pod.Namespace, secretRef.Name)
 		if err != nil {
 			klog.InfoS("Unable to retrieve pull secret, the image pull may not succeed.", "pod", klog.KObj(pod), "secret", klog.KObj(secret), "err", err)
```
Comment (SergeyKanzhelev) [inline pkg/kubelet/kubelet_pods.go:894]: btw, this log message tries to output the result of the method execution that returned an `err`.  Perhaps `secretRef.Name` can be logged instead or alongside the `secret` object.

File: pkg/kubelet/kubelet_pods.go (hunk 3 above)
Comment (SergeyKanzhelev) [reply pkg/kubelet/kubelet_pods.go:895]: makes sense.

**Pull request closed without merging**